src/metabase/public_settings/premium_features.clj

```diff
@@ -45,17 +46,27 @@
 
 (declare premium-embedding-token)
 
-(defn- active-users-count* []
-  {:post [(integer? %)]}
-  "Returns the number of active users."
-  (assert ((requiring-resolve 'metabase.db/db-is-set-up?)) "Metabase DB is not yet set up")
-  (t2/count :core_user :is_active true))
-
-(def ^:private cached-active-user-count
-  "Primarily used for the settings because we don't wish it to be 100%."
-  (memoize/ttl
-    active-users-count*
-    :ttl/threshold (u/minutes->ms 5)))
+;; let's prevent the DB from getting slammed with calls to `active-user-count`, we only really need one in flight at a
+;; time.
+(let [f        (fn []
+                 {:post [(integer? %)]}
+                 (log/info (u/colorize :yellow "GETTING ACTIVE USER COUNT!"))
+                 (assert ((requiring-resolve 'metabase.db/db-is-set-up?)) "Metabase DB is not yet set up")
+                 ;; force this to use a new Connection, it seems to be getting called in situations where the Connection
+                 ;; is from a different thread and is invalid by the time we get to use it
+                 (let [result (binding [t2.conn/*current-connectable* nil]
+                                (t2/count :core_user :is_active true))]
+                   (log/info (u/colorize :green "=>") result)
+                   result))
+      memoized (memoize/ttl
+                f
+                :ttl/threshold (u/minutes->ms 5))
+      lock     (Object.)]
+  (defn- cached-active-users-count
+    "Primarily used for the settings because we don't wish it to be 100%. (HUH?)"
+    []
+    (locking lock
```

Comment:
Key thing is putting the lock around the memoized fn call, so we don't have like 10 different threads all hitting it at the same time, and also having the memoized version be the one used elsewhere in this namespace.

Reply:
Could you help me understand how this keeps this function from being called from multiple threads at the same time?

isn't the lock a local object to each thread?